include/flang/Common/idioms.h

```diff
@@ -60,13 +63,16 @@ template<typename... LAMBDAS> struct visitors : LAMBDAS... {
 
 template<typename... LAMBDAS> visitors(LAMBDAS... x)->visitors<LAMBDAS...>;
 
-// Calls std::fprintf(stderr, ...), then abort().
-[[noreturn]] void die(const char *, ...);
-
-#define DIE(x) Fortran::common::die(x " at " __FILE__ "(%d)", __LINE__)
+template <typename... Args> 
+[[noreturn]] void die(const char* c, Args ... args) {
+   std::string buf;
+   llvm::raw_string_ostream ss {buf};
+   ss << llvm::format(c, args...);
+   llvm_unreachable(ss.str().c_str());
```

Comment:
What file and line number is going to be reported here?

Reply:
Probably the file and line number of the header, I'm not sure what the right solution to pass the original file and line number forward would be.

Reply:
As David responded. I overlooked the file name and line number. The file and line number will be always of idioms.h header.
Probably we can retain earlier implementation of die for cases with varargs (with format specifiers).

Reply:
I agree that that's the best thing to do for now. We can always come back to this later and have another look.

Reply:
Alternatively, we use the current DIE macro but call this new die function instead of the old implementation? That should also work?

Reply:
With current implementation, DIE macro and all its usage has been replaced with llvm_unreachable. 
Old implementation of die is required in 8 places. We can leave the function and its usage untouched.